Fix #118: Introduce TopicController

BenHenning opened the pull request:
This introduces the interfaces for TopicController, which is also relevant for the #137 since this is providing the source information for both topics & stories. Note that thumbnails for topics & chapters are solved here (stories do not have thumbnails in these views, but do in others: #176). Thumbnails for skills are intentionally left absent, and I don't expect us to have skill thumbnails for the prototype. We should implement a reasonable default for these views in the meantime.

This includes tests to force test replacement once #15 is resolved. This PR includes no real implementation.

Reference page mocks used to determine what information should be provided by this controller: [topic page](https://xd.adobe.com/view/bee1fced-33bd-4f69-6a89-2c30fe1c4f65-c8d0/screen/94a5f38c-e475-4514-8784-0e42cf70c25e/Home-Page-1-Start-Learning-Pre-Scroll-5), [story page](https://xd.adobe.com/view/bee1fced-33bd-4f69-6a89-2c30fe1c4f65-c8d0/screen/5abb0ace-19ee-48b8-8154-919154f924d2/Home-Page-1-Story-Pre-Scroll-1), and [concept cards](https://xd.adobe.com/view/bee1fced-33bd-4f69-6a89-2c30fe1c4f65-c8d0/screen/5d907b23-a75e-4f8b-8c59-afd3c734c53e/Review-Q1).

This is blocked on #175 hence the base branch.

BenHenning committed:
Introduce TopicController & tests.

BenHenning commented:
I included most of the team in the review as an FYI for how we can structure domain controllers ahead of implementation. This is a good example of that.

I'm happy to submit this with just 1 approval since it's not introducing any logic yet.

BenHenning committed:
Introduce 6 sample thumbnail graphics locally and a proto data structure
for thumbnails that can be used for lessons.

BenHenning committed:
Merge branch 'introduce-initial-lesson-thumbnails-and-structure' into introduce-topic-controller-interface

BenHenning committed:
Update topic data structures to include comments and better thumbnails.

rt4914 commented on domain/src/main/java/org/oppia/domain/topic/TopicController.kt:
```diff
@@ -0,0 +1,212 @@
+import org.oppia.util.data.AsyncResult
+import java.lang.IllegalArgumentException
+import javax.inject.Inject
+import javax.inject.Singleton
```
Alphabetical order for imports.

BenHenning replied:
Done.

rt4914 commented on model/src/main/proto/topic.proto:
```diff
@@ -0,0 +1,83 @@
+  string description = 2;
+
+  // A URL corresponding to a thumbnail associated with this skill. If this is absent, the UI should show a reasonable
+  // alternative.
```
Nit: White space before `alternative` (same from PR #175 )

BenHenning replied:
Per my comment on #175, since this isn't a TODO we should use regular single whitespace continuation like other single-line comments.

rt4914 reviewed (approved):
LGTM. Suggested nit changes.

nikitamarysolomanpvt commented on domain/src/test/java/org/oppia/domain/topic/TopicControllerTest.kt:
```diff
@@ -0,0 +1,278 @@
+    assertThat(getSkillIds(topic)).containsExactly("test_skill_id_0", "test_skill_id_1").inOrder()
+  }
+
+  // This test is intentionally verifying that there are no thumbnails for skills, since there are not yet any to
```
Are we supposed to do multiline comments like this ?

BenHenning replied:
Multi-line comments can be done like this, or as a block comment, e.g.:

```kotlin
/*
 * First line
 * Second line.
 */
```

Or more compactly:

```kotlin
/* First line.
 * Second line. */
```

Note the single '*' in the beginning to indicate this isn't a Javadoc.

In this case, I'm just providing context on why this test is set up in this way. This could be framed as either single-line comments or a documentation comment, though your point makes me think it might be better as an in-code comment instead of a method-level comment.

nikitamarysolomanpvt commented:
LGTM

jamesxu0 commented:
Add method to get concept card data

BenHenning committed:
Merge branch 'develop' into introduce-topic-controller-interface

Conflicts:
	model/src/main/proto/topic.proto

BenHenning commented:
Thanks for the reminder @jamesxu0, I didn't quite get to that yet. Keeping this open until I resolve that bit.

BenHenning committed:
Address review comments.

BenHenning committed:
Introduce support for retrieving concept cards in TopicController (in a
stubbed way), without tests.

This also cleans up the voiceover/translation proto dependencies a bit,
and migrates Playability over to the ChapterPlayState enum.

BenHenning commented:
@jamesxu0 PTAL. I added concept card support, but ran out of time before I could add tests. Hopefully this is sufficient to unblock you if you branch off of this PR's branch (please let me know if that isn't the case).

jamesxu0 reviewed (approved):
LGTM, this should unblock me for concept card, thanks!

BenHenning committed:
Merge branch 'develop' into introduce-topic-controller-interface

Conflicts:
	model/src/main/proto/topic.proto

BenHenning committed:
Add remaining tests. Fix post-merge build. Clean up proto files.

BenHenning commented:
@jamesxu0 I just added tests for getConceptCard in TopicControllerTest. I'd like if you could take a quick glance at them to make sure everything looks good, though I'm fine submitting this as-is.

jamesxu0 reviewed (approved):
Concept Card tests look good! LGTM

BenHenning commented:
Thanks!
